kernel/comps/framebuffer/src/ansi_escape.rs

```diff
@@ -152,26 +152,99 @@ impl EscapeFsm {
+                    cursor += 2;
+                }
+                38 if num_params - cursor >= 4 && self.params[cursor] == 2 => {
+                    op.set_fg_color(Pixel {
+                        red: self.params[cursor + 1] as u8,
+                        green: self.params[cursor + 2] as u8,
+                        blue: self.params[cursor + 3] as u8,
+                    });
+                    cursor += 4;
+                }
+                // Reset to the default foreground color
+                39 => op.set_fg_color(Pixel::WHITE),
+                90..=97 => op.set_fg_color(COLORS[op_code as usize - 90 + 8]),
+
+                // Set background colors
+                // Reference: <https://en.wikipedia.org/wiki/ANSI_escape_code#Colors>
+                40..=47 => op.set_bg_color(COLORS[op_code as usize - 40]),
+                48 if num_params - cursor >= 2 && self.params[cursor] == 5 => {
+                    op.set_bg_color(Self::get_256_color(self.params[cursor + 1] as u16));
+                    cursor += 2;
+                }
+                48 if num_params - cursor >= 4 && self.params[cursor] == 2 => {
+                    op.set_bg_color(Pixel {
+                        red: self.params[cursor + 1] as u8,
+                        green: self.params[cursor + 2] as u8,
+                        blue: self.params[cursor + 3] as u8,
+                    });
+                    cursor += 4;
+                }
+                // Reset to the default background color
+                49 => op.set_bg_color(Pixel::BLACK),
+                100..=107 => op.set_bg_color(COLORS[op_code as usize - 100 + 8]),
 
                 // Invalid or unsupported
-                _ => {}
+                _ => return,
             }
         }
     }
+
+    /// Gets the 256-color used by Linux TTY.
+    ///
+    /// Reference: <https://elixir.bootlin.com/linux/v6.16.7/source/drivers/tty/vt/vt.c#L1599>
+    fn get_256_color(mut i: u16) -> Pixel {
+        let (red, green, blue) = match i {
+            // Standard colors.
+            0..=7 => {
+                let r = if i & 1 != 0 { 0xaa } else { 0x00 };
+                let g = if i & 2 != 0 { 0xaa } else { 0x00 };
+                let b = if i & 4 != 0 { 0xaa } else { 0x00 };
+                (r, g, b)
+            }
+            8..=15 => {
+                let r = if i & 1 != 0 { 0xff } else { 0x55 };
+                let g = if i & 2 != 0 { 0xff } else { 0x55 };
+                let b = if i & 4 != 0 { 0xff } else { 0x55 };
+                (r, g, b)
+            }
+            // 6x6x6 color cube.
+            16..=231 => {
+                i -= 16;
+                let b = i % 6 * 255 / 6;
+                i /= 6;
+                let g = i % 6 * 255 / 6;
+                i /= 6;
+                let r = i * 255 / 6;
+                (r as u8, g as u8, b as u8)
+            }
+            // Grayscale ramp.
+            _ => {
+                let g = (i * 10 - 2312) as u8;
```

Comment:
This will cause panic in debug builds if the user tries to give us an invalid large value, right? I don't think the result matters if the user provides an invalid value, but we need to ensure that the kernel does not panic. The kernel shouldn't panic, even in debug builds.

Maybe `get_256_color` should still accept a `u8`. It can internally use `u16`, like this:
```rust
    fn get_256_color(i: u8) -> Pixel {
        // All arithmetic operations below won't overflow because `i` comes from a `u8`.
        let mut i = i as u16;
```

Or we can also just change this line to `i.wrapping_mul(10).wrapping_sub(2312) as u8`. It looks like a bit odd, though.